diffrax/_progress_bar.py

```diff
@@ -0,0 +1,186 @@
+import abc
+from typing import Generic, NamedTuple, TypeVar
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax.experimental import io_callback
+
+from ._custom_types import FloatScalarLike, IntScalarLike
+
+
+ProgressBarGenericState = TypeVar("ProgressBarGenericState")
+
+
+class AbstractProgressBar(eqx.Module, Generic[ProgressBarGenericState]):
+    @abc.abstractmethod
+    def init(self) -> ProgressBarGenericState:
+        """Creates a new progress bar and returns its unique identifier"""
+
+    @abc.abstractmethod
+    def step(
+        self, state: ProgressBarGenericState, progress: FloatScalarLike
+    ) -> ProgressBarGenericState:
+        """Updates the progress bar and returns the last printed progress value"""
+
+    @abc.abstractmethod
+    def close(self, state: ProgressBarGenericState):
+        """Closes the progress bar"""
+
+
+NoProgressBarState = NamedTuple("NoProgressBarState", [])
+
+
+class NoProgressBar(AbstractProgressBar):
+    def init(self) -> NoProgressBarState:
+        return NoProgressBarState()
+
+    def step(
+        self,
+        state: NoProgressBarState,
+        progress: FloatScalarLike,
+    ) -> NoProgressBarState:
+        return state
+
+    def close(self, state: NoProgressBarState):
+        pass
+
+
+_TextProgressBarState = NamedTuple(
+    "_TextProgressBarState", [("progress", FloatScalarLike)]
+)
+
+
+class TextProgressBar(AbstractProgressBar):
+    minimum_increase: float
+
+    def __init__(self, minimum_increase: float = 2.0):
+        """
+        **Arguments:**
+            - `minimum_increase`: minimum increase of progress bar in percent
+        """
+        self.minimum_increase = minimum_increase
+
+    def init(self) -> _TextProgressBarState:
+        return _TextProgressBarState(progress=0.0)
+
+    def step(
+        self,
+        state: _TextProgressBarState,
+        progress: FloatScalarLike,
+    ) -> _TextProgressBarState:
+        """
+        **Arguments:**
+            - `state`: progress bar state
+            - `progress`: current progress in percent
+        """
```

Comment:
I think this docstring can be removed as it's already covered in the abstract base class.